libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -413,39 +399,43 @@ method publish*(g: GossipSub,
 
   trace "about to publish message on topic", name = topic,
                                              data = data.shortLog
+
+  var peers: HashSet[string]
+  var tries = 5 # try publishing up to 5 times
   if data.len > 0 and topic.len > 0:
-    var peers: HashSet[string]
-    if topic in g.topics: # if we're subscribed to the topic attempt to build a mesh
-      await g.rebalanceMesh(topic)
-      peers = g.mesh[topic]
-    else: # send to fanout peers
-      await g.replenishFanout(topic)
-      if topic in g.fanout:
-        peers = g.fanout[topic]
-        # set the fanout expiery time
-        g.lastFanoutPubSub[topic] = Moment.fromNow(GossipSubFanoutTTL)
-
-    let msg = newMessage(g.peerInfo, data, topic, g.sign)
-    var sent: seq[Future[void]]
-    for p in peers:
-      if p == g.peerInfo.id:
-        continue
+    while peers.len <= 0 and tries > 0:
+      if topic in g.topics: # if we're subscribed to the topic attempt to build a mesh
+        await g.rebalanceMesh(topic)
+        peers = g.mesh[topic]
+      else: # send to fanout peers
+        await g.replenishFanout(topic)
+        if topic in g.fanout:
+          peers = g.fanout[topic]
+          # set the fanout expiry time
+          g.lastFanoutPubSub[topic] = Moment.fromNow(GossipSubFanoutTTL)
+
+      let msg = newMessage(g.peerInfo, data, topic, g.sign)
+      var sent: seq[Future[void]]
+      for p in peers:
+        if p == g.peerInfo.id:
+          continue
+
+        trace "publishing on topic", name = topic
+        g.mcache.put(msg)
+        sent.add(g.peers[p].send(@[RPCMsg(messages: @[msg])]))
```

Comment:
try/catch?

Reply:
Tend wont throw right now and the futures are accumulated and then awaited with `allFutures` which won't throw either. Tho, I've reworked that part a bit.

Reply:
We removed `allFutures` from the codebase sometime ago, if possible stick with `allFinished` and the macro

Reply:
its `allFinished`, typo.